### ISSUE TITLE
Consider resurrecting the mingw_pread patches

### ISSUE THREAD

**Issue by sschuberth** [git-for-windows/git#94]
Chromium uses a [Windows pread performance patch](https://chromium-review.googlesource.com/#/c/186104/) since quite a while, and Stefan Zager had proposed an [index-pack threading patch](http://git.661346.n2.nabble.com/PATCH-Enable-index-pack-threading-in-msysgit-td7606049.html) on top of it.

I believe it makes sense to look at integrating this again, and this issue exists so we do not forget about it.

@kblees Would you have time to look at the patch again?

**Comment by dscho** [comment]
According to https://msdn.microsoft.com/en-us/library/windows/desktop/aa365497%28v=vs.85%29.aspx, the `ReOpenFile` function is only available starting with Vista, and according to [these](https://github.com/git-for-windows/git/issues/57) two [issues](https://github.com/git-for-windows/git/issues/52), we still have a contributor who uses XP himself...

**Comment by dscho** [comment]
For reference, I cherry-picked Zach's patch: https://github.com/git-for-windows/git/compare/master...pread

**Comment by dscho** [comment]
I have no physical setup right now to do proper performance tests, but we would really need to do them *first*. And then, *if* it is worth it, we can always play games with lazy-loading the `ReOpenFile` symbol and falling back to [the `malloc`-based emulation](https://github.com/git-for-windows/git/blob/master/compat/pread.c).

**Comment by kblees** [comment]
The NO_THREAD_SAFE_PREAD setting is gone from the Makefile, thus I suspect the problem has already been fixed upstream (e.g. by using Duy's patch [1]).

For reference, upstream discussion is at \[2] (gmane seems to be lacking the concluding messages). I also provided a thread-safe mingw_pread based on `ReadFile` + `OVERLAPPED` that doesn't require Windows Vista [3].

[1] http://article.gmane.org/gmane.comp.version-control.git/196042
[2] http://git.661346.n2.nabble.com/Make-the-git-codebase-thread-safe-td7603504.html
[3] http://article.gmane.org/gmane.comp.version-control.git/242120

**Comment by kblees** [comment]
...indeed: 39539495 "index-pack: work around thread-unsafe pread()" merged on 2014-03-25 (53f52cd9)

**Comment by dscho** [comment]
@kblees but we still define `NO_PREAD = YesPlease`, should we stop that?

**Comment by sschuberth** [comment]
> but we still define NO_PREAD = YesPlease, should we stop that?

I don't think we can stop that yet as there's still no `mingw_pread()` (neither Karsten's nor Stefan's) in our code as far as I can see.

> For reference, I cherry-picked Zach's patch

Who's Zach? :-) Also, we probably would want to fix the indentation in that patch before merging it.

**Comment by kblees** [comment]
> but we still define NO_PREAD = YesPlease, should we stop that?

No, we still need NO_PREAD to get `git_pread()` from compat/pread.c.

We had this discussion last year because the multi-threaded index-pack algorithm required a thread-safe `pread()` at the time, and `git_pread()` is not thread-safe.

One possible solution was to provide a thread-safe `mingw_pread()`...until Duy proposed to change the algorithm so that it no longer requires a thread-safe implementation.

In other words: the `git_pread()` implentation we get with NO_PREAD is perfectly fine.

**Comment by sschuberth** [comment]
And we also would not gain any performance in some other area by still providing a thread-safe `mingw_pread()`, correct?

**Comment by kblees** [comment]
> And we also would not gain any performance in some other area by still providing a thread-safe mingw_pread(), correct?

I don't think so. My version does essentially the same thing as `git_pread()` (just using Win32 `ReadFile()` instead of `read()`). I would think that the `ReOpenFile()` variant is even slower (due to the performance penalty of opening a new file handle on each call).
